examples/cgroup-rstat-flushing.bpf.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Measure cgroup rstat (recursive stats) flushing overhead and latency.
+ *
+ * Loosely based on bpftrace script cgroup_rstat_tracepoint.bt
+ *  - https://github.com/xdp-project/xdp-project/blob/master/areas/latency/cgroup_rstat_tracepoint.bt
+ *
+ * Depends on tracepoints added in kernel v6.10
+ */
+
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "maps.bpf.h"
+
+#define MAX_CGRP_LEVELS	5
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, MAX_CGRP_LEVELS + 1);
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_flush_total SEC(".maps");
+
+/* Total counter for obtaining lock together with contended state
+ *
+ * This counter also contains "yield" case. To determine "normal" lock
+ * case subtract "yield" counter in prometheus query.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 2); /* contended state used as key */
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_locked_total SEC(".maps");
+
+/* Counter for obtaining lock again after yield (and contended state).
+ *
+ * Kernel can yield the rstat lock when walking individial CPU stats.
+ * This leads to "interresting" concurrency issues.  Thus, keep
+ * seperate counter for "yield" cases to help diagnose.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 2); /* contended state used as key */
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_locked_yield SEC(".maps");
+
+/* Counter for lock contended case, recorded per cgroup level */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, MAX_CGRP_LEVELS + 1);
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_lock_contended SEC(".maps");
+
+/** Measurement#1: lock rates
+ *  =========================
+ For locks, the problematic/interesting case is when the lock was contended.
+
+ Simply counting the lock+unlock is complicated by yielding the lock in the main
+ flushing loop (in cgroup_rstat_flush_locked()). The tracepoints "cpu" argument
+ will be (minus) -1 when the lock is not a yielded lock.
+
+ Concern: The lock rates will vary (a lot) and aggregating this as an average
+ will not capture spikes. Especially given Prometheus capture intervals only
+ happens every 53 seconds.
+
+ Q: will lock "type=yield" be a good idea?
+
+*/
+
+SEC("tp_btf/cgroup_rstat_locked")
```

Comment:
This is too new for the Ubuntu we run in CI, so you need to add an exclude to `CONFIGS_TO_IGNORE_IN_CHECK` in the top level `Makefile`.

Reply:
Makes sense.
I'll do that.